Tests/src/InternalTests/SpaceEntityTests.cpp

```diff
@@ -0,0 +1,1093 @@
+{
+	SetRandSeed();
+
+	auto& SystemsManager = csp::systems::SystemsManager::Get();
+	auto* UserSystem	 = SystemsManager.GetUserSystem();
+	auto* SpaceSystem	 = SystemsManager.GetSpaceSystem();
+	auto* Connection	 = SystemsManager.GetMultiplayerConnection();
+	auto* EntitySystem	 = SystemsManager.GetSpaceEntitySystem();
+
+	const char* TestSpaceName		 = "OLY-UNITTEST-SPACE-REWIND";
+	const char* TestSpaceDescription = "OLY-UNITTEST-SPACEDESC-REWIND";
+
+	char UniqueSpaceName[256];
+	SPRINTF(UniqueSpaceName, "%s-%s", TestSpaceName, GetUniqueString().c_str());
+
+	csp::common::String UserId;
+
+	// Log in
+	LogIn(UserSystem, UserId);
+
+	// Create space
+	csp::systems::Space Space;
+	CreateSpace(SpaceSystem,
+				UniqueSpaceName,
+				TestSpaceDescription,
+				csp::systems::SpaceAttributes::Private,
+				nullptr,
+				nullptr,
+				nullptr,
+				nullptr,
+				Space);
+
+	std::atomic_bool ScriptSystemReady = false;
+
+	auto ScriptSystemReadyCallback = [&ScriptSystemReady](bool Ok)
+	{
+		EXPECT_EQ(Ok, true);
+		std::cerr << "ScriptSystemReadyCallback called" << std::endl;
+		ScriptSystemReady = true;
+	};
+
+	auto [EnterResult] = AWAIT_PRE(SpaceSystem, EnterSpace, RequestPredicate, Space.Id);
+
+	EXPECT_EQ(EnterResult.GetResultCode(), csp::systems::EResultCode::Success);
+
+	EntitySystem->SetEntityCreatedCallback(
+		[](SpaceEntity* Entity)
+		{
+		});
+	EntitySystem->SetScriptSystemReadyCallback(ScriptSystemReadyCallback);
+
+	const csp::common::String& UserName = "Player 1";
+	const SpaceTransform& UserTransform
+		= {csp::common::Vector3 {1.452322f, 2.34f, 3.45f}, csp::common::Vector4 {4.1f, 5.1f, 6.1f, 7.1f}, csp::common::Vector3 {1, 1, 1}};
+	AvatarState UserAvatarState				  = AvatarState::Idle;
+	const csp::common::String& UserAvatarId	  = "MyCoolAvatar";
+	AvatarPlayMode UserAvatarPlayMode		  = AvatarPlayMode::Default;
+	LocomotionModel UserAvatarLocomotionModel = LocomotionModel::Grounded;
+
+	auto [Avatar] = AWAIT(EntitySystem, CreateAvatar, UserName, UserTransform, UserAvatarState, UserAvatarId, UserAvatarPlayMode);
+	EXPECT_NE(Avatar, nullptr);
+
+	std::cerr << "CreateAvatar Local Callback" << std::endl;
+
+	EXPECT_EQ(Avatar->GetEntityType(), SpaceEntityType::Avatar);
+
+	if (Avatar->GetEntityType() == SpaceEntityType::Avatar)
+	{
+		OnUserCreated(Avatar);
+	}
+
+	const std::string ScriptText = R"xx(
+		
+		var entities = TheEntitySystem.getEntities();
+		var entityIndex = TheEntitySystem.getIndexOfEntity(ThisEntity.id);
+
+		globalThis.onTick = () => {
+			var entity = entities[entityIndex];
+			var parent = TheEntitySystem.getEntityByName("Object 1");
+
+			entity.rotation = parent.globalRotation;
+		}
+
+		ThisEntity.subscribeToMessage("entityTick", "onTick");
+
+	)xx";
+
+	auto ScriptSystemIsReady = [&ScriptSystemReady]()
+	{
+		std::cerr << "Waiting for ScriptSystemReady" << std::endl;
+		return (ScriptSystemReady == true);
+	};
+
+	EXPECT_EQ(ResponseWaiter::WaitFor(ScriptSystemIsReady, std::chrono::seconds(5)), true);
+
+	{
```

Comment:
[Question] Why the additional scope here, is there some RAII thing I'm missing?

Reply:
These test are based of the existing script system test, the scoping is just to separate the main test from the require setup and destruction step such as logging in, ensuring the script system is available, etc

Reply:
Just organizational then? Fair enough.